ee/clickhouse/models/property.py

```diff
@@ -230,25 +232,45 @@ def prop_filter_json_extract(
             ),
             params,
         )
+    elif operator == "is_date_exact":
+        # TODO introducing duplication in these branches now rather than refactor too early
+        assert isinstance(prop.value, str)
+        prop_value_param_key = "v{}_{}".format(prepend, idx)
+
+        # if we're comparing against a date with no time,
+        # truncate the values in the DB which may have times
+        granularity = "day" if re.match(r"^\d{4}-\d{2}-\d{2}$", prop.value) else "second"
+        query = f"""AND date_trunc('{granularity}', coalesce(
+            parseDateTimeBestEffortOrNull({property_expr}),
+            parseDateTimeBestEffortOrNull(substring({property_expr}, 1, 10))
+        )) = %({prop_value_param_key})s"""
+
+        return (
+            query,
+            {"k{}_{}".format(prepend, idx): prop.key, prop_value_param_key: prop.value,},
+        )
     elif operator == "is_date_after":
-        # introducing duplication in these branches now rather than refactor too early
+        # TODO introducing duplication in these branches now rather than refactor too early
         assert isinstance(prop.value, str)
         prop_value_param_key = "v{}_{}".format(prepend, idx)
 
-        query = f"""AND coalesce(
-                parseDateTimeBestEffortOrNull({property_expr}),
-                parseDateTimeBestEffortOrNull(substring({property_expr}, 1, 10))
-            ) > %({prop_value_param_key})s"""
+        # if we're comparing against a date with no time,
+        # then instead of 2019-01-01 (implied 00:00:00)
+        # use 2019-01-01 23:59:59
+        is_date_only = re.match(r"^\d{4}-\d{2}-\d{2}$", prop.value)
+        adjustment = 1 if is_date_only else 0
+        granularity = "day" if is_date_only else "second"
+        query = f"""AND date_trunc('{granularity}', subtractSeconds(addDays(coalesce(
```

Comment:
What was the precision in ClickHouse again? Just double checking that there's no way 23:59:59.499 or some other date in a property that contains milliseconds somehow gets included in the range when it shouldn't be. I think an edge case like that should get a test as well.

Reply:
Well, precision in clickhouse is unknown, people can store any format of DateTime we recognise. Some of our SDK provided values have milliseconds.

Precision in the UI is limited to seconds by ANTd's time-picker

ClickHouse doesn't provide a `subtractMilliseconds` function https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions/#subtractyears-subtractmonths-subtractweeks-subtractdays-subtracthours-subtractminutes-subtractseconds-subtractquarters

So, I think, that means that anything between 59 seconds and the next minute is counted as before the hour. My understanding is that the date_trunc function acts as a floor, not a round

```
select
       parseDateTimeBestEffort('2015-08-29T11:22:09.815479Z'),
       date_trunc('second', parseDateTimeBestEffort('2015-08-29T11:22:09.815479Z'))
```

<img width="810" alt="Screenshot 2022-02-01 at 09 44 32" src="https://user-images.githubusercontent.com/984817/151945865-ddf55daa-53f4-47e6-b7dc-50bd15cd09f3.png">

I'll add the test case but based on that am confident that ClickHouse is operating at second precision not millisecond precision

Reply:
Added the test case...

Let's pretend that didn't expose that the query wasn't right, and never talk about this again 😅

Reply:
> the datepicker closing itself when I select a date (my intention was to next click "include time") threw me off:

Yep, ANT design doesn't present the OK button for a Date picker, only for a DateTime picker.

https://github.com/ant-design/ant-design/issues/22966